File: lang/go/cxlib/atom_test.go
package cxlib

import (
	"strings"
	"testing"
)

// atom_test.go Tier-1 binding catchup for Go.
//
// Covers all six atom-scalar-kind gates:
//
//   33.1 — `:NAME` parses through libcx and decodes into the typed
//          AtomValue wrapper at the Go binding layer.
//   33.2 — Atom equality is type-strict (AtomValue != string with the
//          same content).
//   33.3 — Atom round-trips through ast_bin via cx_to_ast_bin →
//          decodeAST → encodeAST → cx_ast_bin_to_cx with byte-identical
//          canonical output.
//   33.4 — Canonical render emits `:NAME` (no quoting) in both attribute
//          and scalar positions.
//   33.5 — Reserved names `:true` / `:false` / `:null` are rejected at
//          construction time by the Go Layer-1 surface.
//   33.6 — Identity-hash domain is disjoint from same-named strings.
//
// These exercise the Go-side surface only; deeper semantic coverage
// (parser disambiguation, evaluator dispatch) lives in V's
// `vcx/cx/atom_test.v` and the cross-binding `conformance_code.go` runner.

// ── Layer-1 surface (gate 33.5) ──────────────────────────────────────────────

func TestAtomConstructorReturnsTypedValue(t *testing.T) {
	a, err := Atom("ok")
	if err != nil {
		t.Fatalf("Atom: %v", err)
	}
	if a.Name != "ok" {
		t.Errorf("Atom.Name = %q, want %q", a.Name, "ok")
	}
}

func TestAtomConstructorAcceptsKebabAndUnderscores(t *testing.T) {
	for _, n := range []string{"not-found", "http_2", "A", "_private"} {
		if _, err := Atom(n); err != nil {
			t.Errorf("Atom(%q) unexpected error: %v", n, err)
		}
	}
}

func TestAtomConstructorRejectsReservedNames(t *testing.T) {
	for _, n := range []string{"true", "false", "null"} {
		_, err := Atom(n)
		if err == nil {
			t.Errorf("Atom(%q) should fail (reserved name)", n)
			continue
		}
		if !strings.Contains(err.Error(), "reserved") {
			t.Errorf("Atom(%q) error = %v, want substring 'reserved'", n, err)
		}
	}
}

func TestAtomConstructorRejectsInvalidIdentifier(t *testing.T) {
	for _, n := range []string{"", "1abc", "has space", "has.dot", "has/slash"} {
		if _, err := Atom(n); err == nil {
			t.Errorf("Atom(%q) should fail (invalid identifier)", n)
		}
	}
}

func TestMustAtomPanicsOnReserved(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Errorf("MustAtom(\"true\") should panic")
		}
	}()
	_ = MustAtom("true")
}

// ── Predicate + accessor (gate 33.2) ─────────────────────────────────────────

func TestIsAtomPredicate(t *testing.T) {
	a := MustAtom("ok")
	if !IsAtom(a) {
		t.Errorf("IsAtom(AtomValue) = false, want true")
	}
	// Type-strict — strings are not atoms.
	if IsAtom("ok") {
		t.Errorf("IsAtom(\"ok\") = true, want false")
	}
	if IsAtom(42) || IsAtom(nil) || IsAtom(true) {
		t.Errorf("IsAtom should reject non-AtomValue types")
	}
}

func TestAtomNameAccessor(t *testing.T) {
	a := MustAtom("not-found")
	if a.AtomName() != "not-found" {
		t.Errorf("AtomValue.AtomName() = %q, want %q", a.AtomName(), "not-found")
	}
	// Function form.
	name, ok := AtomName(a)
	if !ok || name != "not-found" {
		t.Errorf("AtomName(AtomValue) = (%q, %v), want (\"not-found\", true)", name, ok)
	}
}

func TestAtomNameRejectsNonAtom(t *testing.T) {
	// Type-strict — no string coercion.
	if name, ok := AtomName("ok"); ok || name != "" {
		t.Errorf("AtomName(\"ok\") = (%q, %v), want (\"\", false)", name, ok)
	}
	if _, ok := AtomName(nil); ok {
		t.Errorf("AtomName(nil) should return ok=false")
	}
}

// ── Equality (gate 33.2) ─────────────────────────────────────────────────────

func TestAtomEqualityTypeStrict(t *testing.T) {
	a := MustAtom("ok")
	b := MustAtom("ok")
	if a != b {
		t.Errorf("two AtomValue('ok') should compare equal by value")
	}
	c := MustAtom("err")
	if a == c {
		t.Errorf("AtomValue('ok') and AtomValue('err') must be unequal")
	}
	// Type-strict — atom vs string interface-eq must be false.
	var anyA any = a
	var anyS any = "ok"
	if anyA == anyS {
		t.Errorf("AtomValue('ok') == \"ok\" violates type-strict equality")
	}
}

// ── String render (gate 33.4) ────────────────────────────────────────────────

func TestAtomStringRender(t *testing.T) {
	a := MustAtom("ok")
	if a.String() != ":ok" {
		t.Errorf("AtomValue.String() = %q, want %q", a.String(), ":ok")
	}
	if MustAtom("not-found").String() != ":not-found" {
		t.Errorf("kebab-name render mismatch")
	}
}

// ── ast_bin round-trip (gate 33.3 + 33.4) ────────────────────────────────────

func TestAtomDecodesToAtomValue(t *testing.T) {
	doc, err := Parse("[event kind=:click]")
	if err != nil {
		t.Fatalf("Parse: %v", err)
	}
	el, ok := doc.Elements[0].(*Element)
	if !ok {
		t.Fatalf("first element type %T, want *Element", doc.Elements[0])
	}
	if len(el.Attrs) != 1 {
		t.Fatalf("attr count = %d, want 1", len(el.Attrs))
	}
	attr := el.Attrs[0]
	if attr.DataType != "atom" {
		t.Errorf("attr.DataType = %q, want \"atom\"", attr.DataType)
	}
	if !IsAtom(attr.Value) {
		t.Errorf("attr.Value type %T, want AtomValue", attr.Value)
	}
	name, ok := AtomName(attr.Value)
	if !ok || name != "click" {
		t.Errorf("AtomName(attr.Value) = (%q, %v), want (\"click\", true)", name, ok)
	}
}

func TestAtomAstBinRoundTrip(t *testing.T) {
	src := "[event kind=:click]"
	doc, err := Parse(src)
	if err != nil {
		t.Fatalf("Parse: %v", err)
	}
	// Document.ToCx hands the document back through the C ABI via
	// encodeAST → cx_ast_bin_to_cx, so this exercises the full round-trip
	// (gate 33.3 + canonical render gate 33.4).
	out := strings.TrimSpace(doc.ToCx())
	if out != src {
		t.Errorf("round-trip mismatch:\n  got:  %q\n  want: %q", out, src)
	}
}

func TestAtomInElementBodyRoundTrip(t *testing.T) {
	// Atom inside an element body (not just attribute).
	src := "[state :idle]"
	doc, err := Parse(src)
	if err != nil {
		t.Fatalf("Parse: %v", err)
	}
	out := strings.TrimSpace(doc.ToCx())
	if out != src {
		t.Errorf("body-position round-trip mismatch:\n  got:  %q\n  want: %q", out, src)
	}
}

// ── Identity-hash disjoint domain (gate 33.6) ───────────────────────────────

func TestAtomHashDisjointFromString(t *testing.T) {
	// libcx's cx_hash over the canonical form is the load-bearing hash;
	// V's atom_test.v covers it directly. Here we mirror the test by
	// confirming canonical forms differ — that propagates through to the
	// SHA-256 over the canonical bytes.
	atomCanon, err := Canonical("[v kind=:ok]")
	if err != nil {
		t.Fatalf("Canonical(atom): %v", err)
	}
	stringCanon, err := Canonical("[v kind='ok']")
	if err != nil {
		t.Fatalf("Canonical(string): %v", err)
	}
	if atomCanon == stringCanon {
		t.Errorf("canonical form collision: atom %q == string %q", atomCanon, stringCanon)
	}

	atomHash, err := Hash("[v kind=:ok]")
	if err != nil {
		t.Fatalf("Hash(atom): %v", err)
	}
	stringHash, err := Hash("[v kind='ok']")
	if err != nil {
		t.Fatalf("Hash(string): %v", err)
	}
	if atomHash == stringHash {
		t.Errorf("identity hash collision: atom %q == string %q (gate 33.6)", atomHash, stringHash)
	}
}

// ── Capability bit advertisement ─────────────────────────────

func TestFeaturesAdvertisesAtomCapBit(t *testing.T) {
	bits := Features()
	if bits&(1<<33) == 0 {
		t.Errorf("cx_features = 0x%x, missing bit 33 (0x200000000) — atom support unadvertised", bits)
	}
}

File: lang/rust/cxlib/tests/atom_test.rs
//! atom scalar kind — Tier-1 binding catchup tests (Rust).
//!
//! Covers all six gates:
//!
//!   33.1 — `:NAME` parses through libcx and decodes into a `Scalar`
//!          node with `data_type=Some("atom")`.
//!   33.2 — Atom equality is type-strict (`Atom` != same-named string).
//!   33.3 — Atom round-trips through ast_bin via `cx_to_ast_bin` →
//!          `binary::decode_ast` → `binary::encode_ast` →
//!          `cx_ast_bin_to_cx` with byte-identical canonical output.
//!   33.4 — Canonical render emits `:NAME` (no quoting) in both
//!          attribute and scalar positions.
//!   33.5 — Reserved names `:true` / `:false` / `:null` are rejected
//!          at construction time by the Rust Layer-1 surface.
//!   33.6 — Identity-hash domain is disjoint from same-named strings.

use cxlib::{atom_name, is_atom, Atom};
use serde_json::Value;

// ── Layer-1 constructor (gates 33.1 + 33.5) ──────────────────────────────────

#[test]
fn atom_new_returns_typed_value() {
    let a = Atom::new("ok").expect("Atom::new('ok')");
    assert_eq!(a.name(), "ok");
    assert_eq!(a, Atom::new("ok").unwrap());
}

#[test]
fn atom_new_accepts_kebab_and_underscores() {
    for n in ["not-found", "http_2", "A", "_private"] {
        assert!(Atom::new(n).is_ok(), "Atom::new({:?}) should succeed", n);
    }
}

#[test]
fn atom_new_rejects_reserved_names() {
    // Gate 33.5 closed-list reservation.
    for n in ["true", "false", "null"] {
        let err = Atom::new(n).unwrap_err();
        assert!(err.contains("reserved"), "Atom::new({:?}) error = {}", n, err);
    }
}

#[test]
fn atom_new_rejects_invalid_identifier() {
    // identifier production `[A-Za-z_][A-Za-z0-9_-]*`.
    for n in ["", "1abc", "has space", "has.dot", "has/slash"] {
        assert!(Atom::new(n).is_err(), "Atom::new({:?}) should fail", n);
    }
}

// ── Predicate + accessor (gate 33.2) ─────────────────────────────────────────

#[test]
fn is_atom_predicate_type_strict() {
    let v = Value::String("ok".to_string());
    assert!(is_atom(Some("atom"), &v));
    // Type-strict — same string content but no atom tag must not match.
    assert!(!is_atom(None, &v));
    assert!(!is_atom(Some("string"), &v));
    // Type-strict — non-string values with the atom tag don't qualify
    // either (the wire form is always Value::String).
    assert!(!is_atom(Some("atom"), &Value::Null));
}

#[test]
fn atom_name_accessor() {
    let v = Value::String("not-found".to_string());
    assert_eq!(atom_name(Some("atom"), &v), Some("not-found"));
    // Type-strict — a same-named string is not coerced.
    assert_eq!(atom_name(None, &v), None);
    assert_eq!(atom_name(Some("string"), &v), None);
}

#[test]
fn atom_method_accessor() {
    let a = Atom::new("ok").unwrap();
    assert_eq!(a.name(), "ok");
}

// ── Equality (gate 33.2) ─────────────────────────────────────────────────────

#[test]
fn atom_equality_type_strict() {
    let a = Atom::new("ok").unwrap();
    let b = Atom::new("ok").unwrap();
    assert_eq!(a, b);
    let c = Atom::new("err").unwrap();
    assert_ne!(a, c);
    // Type-strict — Atom cannot equal a same-named String via Rust's
    // type system (different types, no PartialEq<String> impl).
    // Compile-time assertion: this code must not compile.
    // (We can't easily test the negative compile case here, but the
    // absence of an `impl PartialEq<String> for Atom` is itself the
    // §D2 guarantee.)
}

// ── String render (gate 33.4) ────────────────────────────────────────────────

#[test]
fn atom_display_renders_with_colon_prefix() {
    let a = Atom::new("ok").unwrap();
    assert_eq!(a.to_string(), ":ok");

    let kebab = Atom::new("not-found").unwrap();
    assert_eq!(kebab.to_string(), ":not-found");
}

#[test]
fn atom_as_node_pair_returns_wire_shape() {
    let a = Atom::new("ok").unwrap();
    let (dt, v) = a.as_node_pair();
    assert_eq!(dt, "atom");
    assert_eq!(v, Value::String("ok".to_string()));
}

// ── ast_bin round-trip (gates 33.3 + 33.4) ───────────────────────────────────

#[test]
fn atom_decodes_to_scalar_with_atom_data_type() {
    let doc = cxlib::parse("[event kind=:click]").expect("parse");
    let elem = match &doc.elements[0] {
        cxlib::ast::Node::Element(e) => e,
        other => panic!("first node type {:?}, want Element", other),
    };
    assert_eq!(elem.attrs.len(), 1);
    let attr = &elem.attrs[0];
    assert_eq!(attr.data_type.as_deref(), Some("atom"));
    assert!(is_atom(attr.data_type.as_deref(), &attr.value));
    assert_eq!(atom_name(attr.data_type.as_deref(), &attr.value), Some("click"));
}

#[test]
fn atom_ast_bin_round_trip() {
    // Gate 33.3 + 33.4 — round-trip through cx_to_ast_bin and
    // cx_ast_bin_to_cx preserves the canonical `:click` form.
    let src = "[event kind=:click]";
    // Use the to_cx wrap which routes through cx_to_cx; that path
    // exercises the C-side round-trip directly.
    let out = cxlib::to_cx(src).expect("to_cx");
    assert_eq!(out.trim(), src);
}

#[test]
fn atom_in_element_body_round_trip() {
    let src = "[state :idle]";
    let out = cxlib::to_cx(src).expect("to_cx");
    assert_eq!(out.trim(), src);
}

// ── Identity-hash disjoint domain (gate 33.6) ───────────────────────────────

#[test]
fn atom_hash_disjoint_from_string() {
    // The load-bearing hash is libcx's cx_hash over the canonical form
    // (V's atom_test.v exercises it directly). Mirror that here by
    // confirming canonical forms differ between :ok and 'ok' — the
    // SHA-256 over the canonical bytes propagates the distinction.
    let atom_canon = cxlib::canonical("[v kind=:ok]").expect("canonical(atom)");
    let string_canon = cxlib::canonical("[v kind='ok']").expect("canonical(string)");
    assert_ne!(
        atom_canon, string_canon,
        "canonical form collision: atom={:?} string={:?}", atom_canon, string_canon
    );

    let atom_hash = cxlib::hash("[v kind=:ok]").expect("hash(atom)");
    let string_hash = cxlib::hash("[v kind='ok']").expect("hash(string)");
    assert_ne!(
        atom_hash, string_hash,
        "identity hash collision: atom={:?} string={:?}", atom_hash, string_hash
    );
}

// ── Capability bit advertisement ─────────────────────────────

#[test]
fn features_advertises_atom_cap_bit() {
    // Gate — cx_features() MUST set bit 33 (0x200000000) when the
    // binding parses, evaluates, renders, and round-trips atoms per
    // the gate matrix.
    let bits = cxlib::features();
    assert!(
        bits & (1u64 << 33) != 0,
        "cx_features = 0x{:x}, missing bit 33 (0x200000000) — atom support unadvertised",
        bits
    );
}
